File: bf.py
from enum import Enum
from typing import NamedTuple
from io import StringIO
from pprint import pprint


class BFToken(Enum):
    MoveRight = '>'
    MoveLeft = '<'
    JumpForward = '['
    JumpBackward = ']'
    Incr = '+'
    Decr = '-'
    ReadStdIn = ','
    PutStdOut = '.'

    @classmethod
    def get(cls, token):
        try:
            return cls(token)
        except ValueError:
            return None

    def __repr__(self):
        return f"{self.__class__.__name__}.{self.name}"


class CollapsedToken(NamedTuple):
    token: BFToken
    value: int


def to_u8(value):
    return value % 255


class MismatchedBrackets(SyntaxError):
    pass

def parse(prog):
    loc = 0
    length = len(prog)
    tokens = []
    bracket_stack = []  # idx

    while loc < length:
        symbol = prog[loc]
        op = BFToken.get(symbol)
        loc += 1

        if op is None:
            continue

        if op == BFToken.JumpBackward:
            if not bracket_stack:
                raise MismatchedBrackets(loc)

            open_idx = bracket_stack.pop()
            opened = tokens[open_idx]

            if opened.token != BFToken.JumpForward or opened.value is not None:
                raise MismatchedBrackets(f"Expected to find empty JumpForward found {opened}")

            tokens[open_idx] = opened._replace(value=len(tokens))

            tokens.append(CollapsedToken(token=op, value=open_idx))

        if op == BFToken.JumpForward:
            bracket_stack.append(len(tokens))
            tokens.append(CollapsedToken(token=op, value=None))

        if op not in {BFToken.JumpBackward, BFToken.JumpForward}:
            count = 1
            while loc < length and symbol == prog[loc]:
                count += 1
                loc += 1

            tokens.append(CollapsedToken(token=op, value=count))

    if bracket_stack:
        raise MismatchedBrackets(bracket_stack.pop())

    return tokens


TAPE_SIZE = 30000


def run(ops, stdin, stdout):
    tape = [0] * TAPE_SIZE
    loc = 0
    ptr = 0
    stdin_iter = iter(stdin)
    prog_length = len(ops)

    while loc < prog_length:
        op = ops[loc]

        if op.token == BFToken.MoveRight:
            move_to = op.value + ptr
            if move_to > TAPE_SIZE - 1:
                move_to %= TAPE_SIZE
            ptr = move_to

        if op.token == BFToken.MoveLeft:
            move_to = ptr - op.value
            if move_to < 0:
                move_to %= TAPE_SIZE
            ptr = move_to

        if op.token == BFToken.Incr:
            tape[ptr] = to_u8(tape[ptr] + op.value)

        if op.token == BFToken.Decr:
            tape[ptr] = to_u8(tape[ptr] - op.value)

        if op.token == BFToken.ReadStdIn:
            for _ in range(op.value):
                tape[ptr] = ord(next(stdin_iter, '\0'))

        if op.token == BFToken.PutStdOut:
            for _ in range(op.value):
                stdout.write(chr(tape[ptr]))

        if op.token == BFToken.JumpForward:
            if tape[ptr] == 0:
                loc = op.value

        if op.token == BFToken.JumpBackward:
            if tape[ptr] != 0:
                loc = op.value

        loc += 1


def bf_to_one_line(prog):
    return "".join(prog.split('\n'))


def run_bf(prog, stdout, stdin=""):
    run(parse(bf_to_one_line(prog)), stdin, stdout)


if __name__ == "__main__":
    import sys

    def main():

        if len(sys.argv) == 1:
            sys.stderr.write("Expected at least one argument, a BrainFuck program and optionally input")
            sys.exit(1)

        try:
            with open(sys.argv[1]) as fh:
                prog = fh.read()
        except OSError:
            sys.stderr.write(f"Cannot read {sys.argv[1]}")

        stdin = sys.argv[2] if len(sys.argv) > 2 else ""
        stdout = StringIO()

        try:
            run_bf(prog, stdout, stdin)
        except:
            stdout.seek(0)
            sys.stdout.write("\n")
            sys.stdout.write("Content so far: ")
            sys.stdout.write(stdout.read())
            sys.stdout.write("\n")
            raise

        stdout.seek(0)
        sys.stdout.write('\n'+ stdout.read() + '\n')

    main()

File: bf.rs
use std::char;
use std::num::Wrapping;

const TAPE_SIZE: usize = 30_000;

#[derive(Debug, Eq, PartialEq, Clone)]
enum BrainFuckToken {
    MoveRight,
    MoveLeft,
    Incr,
    Decr,
    Output,
    Input,
    JumpForward,
    JumpBackward
}

impl BrainFuckToken {
    fn from_char(c: char) -> Option<Self> {
         match c {
            '>' => Some(BrainFuckToken::MoveRight),
            '<' => Some(BrainFuckToken::MoveLeft),
            '+' => Some(BrainFuckToken::Incr),
            '-' => Some(BrainFuckToken::Decr),
            '.' => Some(BrainFuckToken::Output),
            ',' => Some(BrainFuckToken::Input),
            '[' => Some(BrainFuckToken::JumpForward),
            ']' => Some(BrainFuckToken::JumpBackward),
            _ => None
        }
    }
}


#[derive(Debug)]
struct Collapsed(BrainFuckToken, usize);

fn lex(prog: String) -> Vec<BrainFuckToken> {
    prog.chars().filter_map(BrainFuckToken::from_char).collect()
}


fn collapse(ops: Vec<BrainFuckToken>) -> Vec<Collapsed> {
    let mut loc: usize = 0;
    let mut collapsed = Vec::with_capacity(ops.len());
    let mut brackets = Vec::new();

    while let Some(symbol) = ops.get(loc) {
        match symbol {
            &BrainFuckToken::JumpForward => {
                brackets.push(collapsed.len());
                collapsed.push(Collapsed(BrainFuckToken::JumpForward, 0));
                loc += 1;
            },

            &BrainFuckToken::JumpBackward => {
                let idx = brackets.pop().unwrap_or_else(|| { panic!("Mismatched brackets: {}", loc); });

                collapsed[idx] = match collapsed.get(idx).unwrap_or_else(|| {
                    panic!("No tokens found at specified stack location: {}", idx);
                })
                {
                    &Collapsed(BrainFuckToken::JumpForward, 0) => {
                        Collapsed(BrainFuckToken::JumpForward, collapsed.len())
                    },
                    &Collapsed(BrainFuckToken::JumpForward, _) => {
                        panic!("Matched populated JumpForward");
                    },
                    tok => {
                        panic!("Matched token was not a JumpForward, got: {:?}", tok);
                    }
                };

                collapsed.push(Collapsed(BrainFuckToken::JumpBackward, idx));

                loc += 1;
            },

            sym @ _ => {
                let mut count = 0;

                loop {
                    match ops.get(loc) {
                        Some(sym) if sym == symbol => {
                            loc += 1;
                            count += 1;
                        },
                        _ => { break; }
                    }
                }

                collapsed.push(Collapsed(sym.clone(), count));
            }
        }
    }

    collapsed.shrink_to_fit();
    collapsed
}


fn run(instructions: Vec<Collapsed>, input: &String) -> String {
    let mut memory = [0u8; TAPE_SIZE];
    let mut instptr: usize = 0;
    let mut memptr: usize = 0;
    let mut input_iter = input.chars();
    let mut output = String::new();

    while let Some(instruction) = instructions.get(instptr) {

        match *instruction {
            Collapsed(BrainFuckToken::MoveRight, x) => {
                memptr = (memptr + x) % TAPE_SIZE;
            },

            Collapsed(BrainFuckToken::MoveLeft, x) => {
                memptr = (((memptr as i32) - (x as i32)) % TAPE_SIZE as i32) as usize;
            },

            Collapsed(BrainFuckToken::Incr, x) => {
                memory[memptr] = (Wrapping::<u8>(memory[memptr]) + Wrapping::<u8>(x as u8)).0;
            },

            Collapsed(BrainFuckToken::Decr, x) => {
                memory[memptr] = (Wrapping::<u8>(memory[memptr]) - Wrapping::<u8>(x as u8)).0;
            },

            Collapsed(BrainFuckToken::Output, x) => {
                let ord = memory[memptr] as u32;
                let chr = char::from_u32(ord).unwrap_or_else(|| panic!("Invalid character: {}", ord));
                for _ in 0..x {
                    output.push(chr);
                }
            },

            Collapsed(BrainFuckToken::Input, x) => {
                for _ in 0..x {
                    memory[memptr] = Wrapping::<u8>(input_iter.next().unwrap_or('\0') as u8).0;
                }
            },

            Collapsed(BrainFuckToken::JumpBackward, ptr) => {
                if memory[memptr] != 0 {
                    instptr = ptr;
                }
            },

            Collapsed(BrainFuckToken::JumpForward, ptr) => {
                if memory[memptr] == 0 {
                    instptr = ptr;
                }
            },
        }
        instptr += 1;
    }

    output
}


fn main() {
    use std::fs::File;
    use std::path::Path;
    use std::io::prelude::*;
    use std::env;

    let arg1 = env::args().nth(1).unwrap();
    let path = Path::new(&arg1);
    let mut s = String::new();
    let mut file = File::open(&path).unwrap();
    file.read_to_string(&mut s).unwrap();
    let input = String::new();

    let lexed = lex(s);
    let tokens = collapse(lexed);
    let result = run(tokens, &input);
    println!("{}", result);
}
